text/0022-gas-spot-market.md

```diff
@@ -0,0 +1,215 @@
+
+The present CEP is intended to guide the development of the gas allocation mechanism that will be present at mainnet launch, 
+later to be supplemented by a gas futures market and a fiat link. The CEP explains the concepts of willingness to pay (WTP), spot gas reservation,
+gas allocation, gas requirements estimation and price determination. Further, it addresses the separation of block gas allocation for regular deploys and WASM-less transfers, 
+balance checks and fallback purses. The present CEP is not intended to be a technical specification, but rather an explanation of the business logic
+and the economic imperatives leading to the proposed deploy lifecycle.
+
+## Motivation
+
+[motivation]: #motivation
+
+The Delta testnet that is active at the time this CEP is being drafted does not have any meaningful gas pricing, although it does charge
+a nominal, hardcoded amount of motes per unit of gas. The reason we need to enable market pricing for gas is simply that gas represents
+time spent on computing the changes to global state, thus making it an approximate measure of a finite resource. This resource is finite
+because meaningful use of a blockchain requires imposition of what is effectively a soft time limit on block execution. Finite resources
+must be priced both because it prevents spam attacks on the blockchain and because it is a means to discover the highest return uses
+for the resource, which ultimately should be allocated to those willing to pay the most.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+In this section, we will outline the basic concepts and UX for blockchain users.
+
+### Basic concepts
+
+#### Willingness to pay (WTP)
+
+Willingness to pay can be seen as analogous to "price" as informally used when discussing user bidding for gas. We will shortly
+address why "price" should not be used for the deploy parameter we are about to define. Willingness to pay is almost self-explanatory, 
+being the amount a user is willing to pay for a unit of gas. We are using the term pragmatically, not intending to mean that this is the
+*maximum* a user is willing to pay. In fact, under our scheme, bidding is almost certainly going to be strategic, with users "shading" their
+true willingness to pay, since the proposed allocation rule is analogous to a first-price auction.
+
+#### Spot gas reservation
+
+Spot gas reservation, conceptually, is the claim a user will lay on gas in a proposed block. Similarly to willingness to pay, it will be
+a deploy parameter. The user effectively buys this amount, priced at the stated willingness to pay, when included in a block.
+
+#### Gas allocation
+
+Gas allocation with a greedy revenue-optimizing ordering of deploys is the core activity that this CEP proposes. It is envisioned that this
+rule will be hardcoded into the block proposer component. The rule will be that highest WTP deploys with the highest spot gas reservations
```

Comment:
That solution may be very far away in the future (year or more), whereas leaving FIFO in place depresses validator revenue and leads to economic inefficiency (someone tinkering around is as likely to get in as someone trying to put through a financially critical transaction). It's also a weak attack vector (with tentative 5000 blocks per week, 10^13 gas limit and price of token at 5 cents, it would cost about $2.5 million to render the platform useless for a week with no op deploys that eat all of the gas).

There's also the slight UI problem of having users specify a gas price that's only ever used for unit conversions. A user paying attention should set this to the minimum of one mote per unit of gas and specify total amount as required to fit the deploy (this behavior is also what causes this to depress validator earnings).